python/dataproc_templates/hive/util/hive_ddl_extractor.py

```diff
@@ -0,0 +1,86 @@
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import argparse
+import pprint
+from logging import Logger
+from typing import Any, Dict, Optional, Sequence
+import dataproc_templates.util.template_constants as constants
+from dataproc_templates import BaseTemplate
+from pyspark.sql import SparkSession
+from datetime import datetime
+
+
+class HiveDDLExtractorTemplate(BaseTemplate): 
+    """
+    Dataproc template implementing exports from Hive to BigQuery
+    """
+
+    @staticmethod
+    def parse_args(args: Optional[Sequence[str]] = None) -> Dict[str, Any]:
+        parser: argparse.ArgumentParser = argparse.ArgumentParser()
+
+        parser.add_argument(
+            f'--{constants.HIVE_DDL_EXTRACTOR_INPUT_DATABASE}',
+            dest=constants.HIVE_DDL_EXTRACTOR_INPUT_DATABASE,
+            required=True,
+            help='Hive database for importing data to BigQuery'
+        )
+
+        parser.add_argument(
+            f'--{constants.HIVE_DDL_EXTRACTOR_OUTPUT_GCS_PATH}',
+            dest=constants.HIVE_DDL_EXTRACTOR_OUTPUT_GCS_PATH,
+            required=True,
+            help='GCS output path'
+        )
+       
+        known_args: argparse.Namespace
+        known_args, _ = parser.parse_known_args(args)
+
+        return vars(known_args)
+
+
+    def run(self, spark: SparkSession, args: Dict[str, Any]) -> None:
+        """
+        
+        Dataproc template allowing the extraction of DDLs from Hive Metastore
+
+        """
+        logger: Logger = self.get_logger(spark=spark)
+
+        hive_database: str = args[constants.HIVE_DDL_EXTRACTOR_INPUT_DATABASE]
+        gcs_output_path: str = args[constants.HIVE_DDL_EXTRACTOR_OUTPUT_GCS_PATH]
+
+        logger.info(
+                "Starting Hive DDL Extraction job with parameters:\n"
+                f"{pprint.pformat(args)}"
+            )
+        
+               
+        def get_ddl(hive_database, table_name):
+            return spark.sql(f"SHOW CREATE TABLE {hive_database}.{table_name}").rdd.map(lambda x: x[0] + ";").collect()[0]
+
+        ct = datetime.now().strftime("%m-%d-%Y %H.%M.%S")
+        output_path = gcs_output_path+"/"+hive_database+"/"+str(ct)
+
+        tables_names = spark.sql(f"SHOW TABLES IN {hive_database}").select("tableName")
```

Comment:
Hey @shubhamgoogle , I changed the code to simplify the arguments, fixed some minor issues, and improved documentation.

Only thing open here is the query, which I removed the "AS SERDE" suffix, because I got this error when testing the template:

`pyspark.sql.utils.AnalysisException: `default`.`empdatanew` is a Spark data source table. Use `SHOW CREATE TABLE` without `AS SERDE` instead`

Could you double check if we really need this?

Reply:
Hi @NiloFreitas , If the table is created by HIVE and we do not use "as serde" using spark this is the output we get:

|CREATE TABLE `default`.`employees` (
  `eid` INT,
  `name` STRING,
  `position` STRING,
  `dept` STRING)
USING text
LOCATION 'hdfs://hive-cluster-m/user/hive/warehouse/partitiondb.db/employee'
TBLPROPERTIES (
  'bucketing_version' = '2',
  'transient_lastDdlTime' = '1658416638')
  
 Now if we use "as serde" in show create command using spark we get:
 
 |CREATE EXTERNAL TABLE `default`.`employees`(
  `eid` INT,
  `name` STRING,
  `position` STRING,
  `dept` STRING)
ROW FORMAT SERDE 'org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe'
**WITH SERDEPROPERTIES (
  'serialization.format' = ',',
  'field.delim' = ',')**
STORED AS
  INPUTFORMAT 'org.apache.hadoop.mapred.TextInputFormat'
  OUTPUTFORMAT 'org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat'
LOCATION 'hdfs://hive-cluster-m/user/hive/warehouse/partitiondb.db/employee'
TBLPROPERTIES (
  'bucketing_version' = '2',
  'transient_lastDdlTime' = '1658416638')
  
As you can see we will not get serde properties like delimiter(which is important for text files) if we do not use "as serde".

"As serde" will not work if the table is created using spark instead, but as it is HIVE DDL extractor I thought we should add it. We can definitely remove it and mention in the document this limitation.

Reply:
Ok. I have put it back and documented. LGTM feel free to merge